pkg/proto/rule/shadow.go

```diff
@@ -0,0 +1,173 @@
+ */
+
+package rule
+
+import (
+	"sync"
+)
+
+import (
+	"github.com/arana-db/arana/pkg/constants"
+)
+
+type (
+	// ShadowRule represents the shadow of databases and tables.
+	ShadowRule struct {
+		mu    sync.RWMutex
+		rules map[string]ShadowRuleManager // map[table]rule
+	}
+
+	Operation struct {
+		enable   bool
+		database string
+		actions  map[string][]*Attribute // map[action][]*Attribute, action in (select, update, delete, update)
+	}
+
+	Attribute struct {
+		column string
+		value  string
+		typ    string // regex, value, hint
+	}
+)
+
+func (o *Operation) GetDatabase() string {
+	return o.database
+}
+
+type ShadowRuleManager interface {
+	MatchValueBy(action, column, value string) bool
+	MatchHintBy(action, hint string) bool
+	MatchRegexBy(action, column, value string) bool
+	GetDatabase() string
+}
+
+func (s *ShadowRule) MatchValueBy(tableName, action, column, value string) bool {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	rule, ok := s.rules[tableName]
+	if !ok {
+		return false
+	}
+
+	return rule.MatchValueBy(action, column, value)
+}
+
+func (s *ShadowRule) MatchHintBy(tableName, action, hint string) bool {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	rule, ok := s.rules[tableName]
+	if !ok {
+		return false
+	}
+	return rule.MatchHintBy(action, hint)
+}
+
+func (s *ShadowRule) MatchRegexBy(tableName, action, column, value string) bool {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	rule, ok := s.rules[tableName]
+	if !ok {
+		return false
+	}
+	return rule.MatchRegexBy(action, column, value)
+}
+
+func (s *ShadowRule) GetDatabase(tableName string) string {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.rules[tableName].GetDatabase()
+}
+
+func (s *ShadowRule) SetRuleManager(tableName string, ruleManager ShadowRuleManager) {
+	s.mu.Lock()
+	if s.rules == nil {
+		s.rules = make(map[string]ShadowRuleManager, 10)
+	}
+	s.rules[tableName] = ruleManager
+	s.mu.Unlock()
+}
+
+func NewRuleManager(actions map[string][]*Attribute, enable bool, database string) ShadowRuleManager {
+	return &Operation{
+		actions:  actions,
+		database: database,
+		enable:   enable,
+	}
+}
+
+func (o *Operation) MatchValueBy(action, column, value string) bool {
+	if !o.enable {
+		return false
+	}
+	attrs, ok := o.actions[action]
+	if !ok {
+		return false
+	}
+
+	for _, attr := range attrs {
+		if attr.typ == constants.ShadowMatchValue && attr.column == column && attr.value == value {
+			return true
+		}
+	}
+	return false
+}
+
+func (o *Operation) MatchHintBy(action, hint string) bool {
+	if !o.enable {
+		return false
+	}
+	attrs, ok := o.actions[action]
+	if !ok {
+		return false
+	}
+
+	for _, attr := range attrs {
+		if attr.typ == constants.ShadowMatchHint {
+			return attr.value == hint
+		}
+	}
+	return false
+}
+
+// MatchRegexBy .
+// TODO impl match regex rule
+func (o *Operation) MatchRegexBy(action, column, value string) bool {
+	if !o.enable {
+		return false
+	}
+	_, ok := o.actions[action]
+	if !ok {
+		return false
+	}
+
+	return false
+}
```

Comment:
```go
_, ok := o.actions[action]
return ok
```

Reply:
Added comments, blanks are for other impl regex match code